utils/statd/hostname.c

```diff
@@ -316,3 +329,57 @@ statd_matchhostname(const char *hostname1, const char *hostname2)
 			(result ? "matched" : "did not match"));
 	return result;
 }
+
+/**
+ * statd_matchhostname_pa - check if two hostnames are equivalent
+ * @hostname1: C string containing hostname
+ * @hostname2: C string containing hostname
+ *
+ * If either hostname is a presentation addresss then further
+ * comparisons beyond strcasecmp are not performed.
+ */
+_Bool
+statd_matchhostname_pa(const char *hostname1, const char *hostname2)
+{
+	struct addrinfo *ai;
+	struct addrinfo hint = {
+		.ai_family	= AF_INET,
+		.ai_flags	= AI_NUMERICHOST,
+		.ai_protocol	= (int)IPPROTO_UDP,
+	};
+	_Bool matched = false;
+
+	if (strcasecmp(hostname1, hostname2) == 0) {
+		matched = true;
+		goto pa_out;
+	}
+
+	/*
+	 * Stop further matching if either hostname is a presentation addresss.
+	 *
+	 * This is used during SM_MON/SM_UNMON requests to avoid stalls from
+	 * flaky DNS servers when locks are requested from clients that have
+	 * provided an ip address for the mon_name.
+	 */
+	ai = get_addrinfo(hostname1, &hint);
+	if (ai != NULL) {
+		/* @hostname1 was a presentation address */
+		freeaddrinfo(ai);
+		goto pa_out;
+	}
+	ai = get_addrinfo(hostname2, &hint);
```

Comment:
If all you're trying to do here is determine if an address is a literal IPv4 address, then a more lightweight way to do that is inet_aton().

Reply:
Great suggestion to use `inet_aton`!